common/main/java/com/couchbase/lite/AbstractReplicator.java

```diff
@@ -382,15 +386,43 @@ public void removeChangeListener(@NonNull ListenerToken token) {
         }
     }
 
-    // This is a workaround: closing at the wrong time (while a push filter is running?) might even
-    // cause a crash.  The issue it addresses is that when a replicator is closed it is detached from
-    // its from LiteCore and receives no further status updates. If it is not already in the STOPPED
-    // state, it will never be in the stopped state.  That means that a database that holds a reference
-    // to it can never close.
-    // Let's just tell the db forget about it.
+    // I've thought a lot about how to implement this.  The problem is that you cannot, fundamentally,
+    // close a replicator(discard its resources before this menthod returns). If it is not stopped,
+    // it will have to be stopped.  I came up with several possibilities.
+    // 1) close it if it is stopped; stop it if it is running.
+    // 2) throw an exception on the attempt to close a running replicator.
+    // 3) tell it to stop and forget it, without verifying that it actually goes away.
+    // 4) Make replicators not Autoclosable.
+    // #4 is very tempting.  Replicators will never be used in a try-with-resources block.
+    // Replicators *do*, however, dominate a LiteCore object that must be freed explicitly.
+    // Making them, transitively, Autoclosable seems correct.
+    // #2 seems to put an excessive burden on client code: it is a bomb that explodes unless you
+    // wait around and pay very close attention to the state of every replicator.
+    // I dislike #1 because it puts the replicator into an odd state in which it is still kinda
+    // half alive.  The client closed it but it isn't quite gone yet.
+    // I have chosen #3.  My rational is that, if you create and start a replicator and then null out
+    // all references to it, this is exactly what will happen: the replicator will be forgotten and
+    // eventually, GCed, told to stop, and freed.  This method does precisely that: best effort to
+    // release the resources and then forget about them.
     public void close() {
-        getDatabase().removeActiveReplicator(this);
+        HashSet<ReplicatorChangeListenerToken> listeners = null;
+        ReplicatorStatus newStatus = null;
+
+        synchronized (getReplicatorLock()) {
+            if (closed) { return; }
+            closed = true;
+
+            if (status.getActivityLevel() != ReplicatorActivityLevel.STOPPED) {
+                listeners = new HashSet<>(changeListeners);
+                newStatus = new ReplicatorStatus(ReplicatorActivityLevel.STOPPED, status.getProgress(), null);
+                status = newStatus;
+            }
+        }
+
         super.close();
+
+        if (newStatus == null) { return; }
+        postChange(true, new ReplicatorChange((Replicator) this, newStatus), listeners);
```

Comment:
I guessed super.close() will just release the LiteCore's replicator. Could this cause a crash if replicator is running? 

I'm not sure about the behavior here when notifying the STOPPED, but the replicator may not be stopped yet? Will it be possible to get notified with BUSY or something else after STOPPED is notified.

Reply:
I have two reasons to believe that it won't:
1) I believe that LiteCore hold references to the replicator.  I believe I just decrement the ref count.  It would be good to confirm this.
2) All replicators are currently released in exactly the way that this close method releases them.  It's been this way since 3.0.  If this way of releasing them might cause crashes, I think we'd have seen them by now.

Reply:
The replicator is, absolutely, stopped, as far as Java is concerned.  It no longer holds any references to the LiteCore replicator object, and the LiteCore object has no valid references to the Java objects.  If the LiteCore replicator attempts to report any status change (or anything like that) to the Java code, the attempt will simply be ignored.